File: game.py
"""
Game controller module for Strands puzzle game.
Manages game flow, word validation, and interaction between components.
"""

from typing import List, Tuple, Optional, Set
from board import Board
from game_state import GameState, WordInfo
from word_finder import WordFinder
from theme import ThemeManager
from ui import StrandsUI


class StrandsGame:
    """Main game controller managing game flow and logic."""
    
    def __init__(self):
        """Initialize the game."""
        # Initialize components
        self.board = Board(rows=6, cols=8)
        self.theme_manager = ThemeManager(self.board)
        self.game_state = GameState(self.board)
        self.word_finder = WordFinder(self.board)
        self.ui = StrandsUI(self.game_state, self.theme_manager)
        
        # Game state tracking
        self.words_found_count = 0
        self.non_theme_words_count = 0
        self.hint_word_count = 0
        
        # Setup callbacks
        self._setup_callbacks()
    
    def _setup_callbacks(self) -> None:
        """Setup UI callbacks."""
        self.ui.on_cell_select = self._on_cell_select
        self.ui.on_word_submit = self._on_word_submit
        self.ui.on_hint_request = self._on_hint_request
    
    def start_new_game(self, theme: Optional[str] = None) -> None:
        """Start a new game with optional theme selection."""
        # Initialize components
        self.board.reset()
        self.game_state.reset()
        self.words_found_count = 0
        self.non_theme_words_count = 0
        self.hint_word_count = 0
        
        # Setup theme
        if theme:
            self.theme_manager.select_theme(theme)
        else:
            self.theme_manager.select_random_theme()
        
        # Generate puzzle
        words, spangram, spangram_path = self.theme_manager.create_puzzle(num_theme_words=5)
        
        # Update game state
        self.game_state.theme_words = set(words)
        self.game_state.game_started = True
        self.game_state.start_time = 0  # Would be set to current time in real implementation
        
        # Initialize word info for theme words
        for word in words:
            self.game_state.found_words[word] = WordInfo(
                word=word,
                path=[],
                found_at=0,
                is_theme=True,
                is_spangram=word == spangram
            )
        
        # Initialize spangram
        if spangram and spangram_path:
            self.game_state.found_words[spangram].path = spangram_path
            self.game_state.found_words[spangram].is_spangram = True
        
        self.ui.reset_game()
    
    def _on_cell_select(self, row: int, col: int) -> None:
        """Handle cell selection."""
        # If cell is already used by another word, don't allow selection
        if self.game_state.is_cell_used(row, col):
            # Clear selection if clicking on used cell
            self.ui.deselect_all()
            return
        
        self.ui.add_cell_to_selection(row, col)
    
    def _on_word_submit(self, word: str) -> bool:
        """Handle word submission."""
        # Validate word length
        if len(word) < 4:
            return False
        
        # Check if word is already found
        if word in self.game_state.found_words_set:
            return False
        
        # Validate path is correct
        path = self.ui.get_current_selection()
        if not self.board.validate_path(path):
            return False
        
        # Check if word matches the selected path
        path_word = self.board.get_letter_path(path)
        if path_word != word:
            return False
        
        # Check if path contains all required cells
        path_set = set(path)
        for row in range(self.board.ROWS):
            for col in range(self.board.COLS):
                if self.board.get_cell(row, col) == word:
                    # Word must contain all occurrences of its letters
                    if (row, col) not in path_set:
                        # Check if the cell is part of a longer word
                        cell_word = self.game_state.get_word_at_cell(row, col)
                        if cell_word != word:
                            return False
        
        # Word is valid - add to game state
        return self._add_found_word(word, path)
    
    def _add_found_word(self, word: str, path: List[Tuple[int, int]]) -> bool:
        """Add a found word to the game state."""
        # Check if word is a spangram
        is_spangram = len(path) == 8 and self._is_spangram_path(path)
        
        # Determine if theme word
        is_theme = word in self.game_state.theme_words
        
        # Update counters
        self.words_found_count += 1
        if not is_theme:
            self.non_theme_words_count += 1
        
        # Add word to game state
        success = self.game_state.add_word(
            word=word,
            path=path,
            is_theme=is_theme,
            is_spangram=is_spangram,
            word_count=self.words_found_count
        )
        
        if success:
            # Check for game completion
            if self.game_state.is_game_complete():
                self.game_state.game_over = True
                self.ui.on_hint_request = lambda: None  # Disable hints on game over
        
        return success
    
    def _is_spangram_path(self, path: List[Tuple[int, int]]) -> bool:
        """Check if path is a spangram (touches two opposite sides)."""
        if len(path) != 8:
            return False
        
        rows = set(p[0] for p in path)
        cols = set(p[1] for p in path)
        
        # Check if touches top and bottom
        if 0 in rows and 5 in rows:
            return True
        
        # Check if touches left and right
        if 0 in cols and 7 in cols:
            return True
        
        return False
    
    def _on_hint_request(self) -> None:
        """Handle hint request."""
        hints_available = self.game_state.get_hints_available()
        if hints_available > 0:
            # Find a word that hasn't been found
            for word in self.theme_manager.current_theme_words:
                if word not in self.game_state.found_words_set:
                    # Show hint for this word
                    print(f"Hint: Try finding '{word}'")
                    return
            
            # If all theme words found, find any word
            all_words = self.word_finder.get_all_words_on_board()
            for word in all_words:
                if word not in self.game_state.found_words_set:
                    print(f"Hint: Try finding '{word}'")
                    return
    
    def get_theme_words_remaining(self) -> List[str]:
        """Get list of theme words not yet found."""
        return self.game_state.get_theme_words_remaining()
    
    def get_theme_words_count(self) -> int:
        """Get count of theme words found."""
        return len(self.game_state.theme_words) - len(self.get_theme_words_remaining())
    
    def get_non_theme_words_count(self) -> int:
        """Get count of non-theme words found."""
        return self.non_theme_words_count
    
    def get_words_found(self) -> List[WordInfo]:
        """Get all found words."""
        return self.game_state.get_found_words()
    
    def is_spangram_found(self) -> bool:
        """Check if spangram has been found."""
        return self.game_state.has_found_spangram()
    
    def get_completion_percentage(self) -> float:
        """Get board completion percentage."""
        return self.game_state.get_completion_percentage()
    
    def get_theme_completion_percentage(self) -> float:
        """Get theme word completion percentage."""
        return self.game_state.get_theme_completion()
    
    def is_game_over(self) -> bool:
        """Check if game is over."""
        return self.game_state.is_game_complete()
    
    def get_score(self) -> int:
        """Get current score."""
        return self.game_state.score
    
    def get_hint_words_remaining(self) -> int:
        """Get remaining hint words needed for next hint."""
        next_hint_threshold = 3 * (self.game_state.hints_used + 1)
        return max(0, next_hint_threshold - self.non_theme_words_count)
    
    def handle_user_action(self, action: str, data: Optional[dict] = None) -> dict:
        """Handle a user action and return response."""
        if action == "select_cell":
            row, col = data["row"], data["col"]
            self.ui.select_cell(row, col)
            return {"success": True, "current_word": self.ui.get_current_word()}
        
        elif action == "submit_word":
            word = data["word"]
            success = self.ui.submit_word()
            return {"success": success, "word": word}
        
        elif action == "clear_selection":
            self.ui.deselect_all()
            return {"success": True, "current_word": ""}
        
        elif action == "hint":
            self.ui.request_hint()
            return {"success": True, "hint_available": self.game_state.get_hints_available() > 0}
        
        elif action == "reset":
            self.start_new_game()
            return {"success": True, "game_reset": True}
        
        elif action == "get_state":
            return {
                "success": True,
                "score": self.get_score(),
                "theme_words_found": self.get_theme_words_count(),
                "theme_words_total": len(self.theme_manager.current_theme_words),
                "non_theme_words_found": self.non_theme_words_count,
                "hints_used": self.game_state.hints_used,
                "completion": self.get_completion_percentage(),
                "game_over": self.is_game_over()
            }
        
        return {"success": False, "error": f"Unknown action: {action}"}

File: game.js
/**
 * Strands Puzzle Game - JavaScript Implementation
 * Client-side game logic and UI handling
 */

// Game configuration
const CONFIG = {
    rows: 6,
    cols: 8,
    themes: {
        fruits: ["APPLE", "BANANA", "CHERRY", "GRAPE", "LEMON"],
        animals: ["LION", "TIGER", "BEAR", "ZEBRA", "GIRAFFE"],
        colors: ["RED", "BLUE", "GREEN", "YELLOW", "ORANGE"],
        countries: ["USA", "CANADA", "MEXICO", "FRANCE", "GERMANY"],
        cities: ["NEW YORK", "LONDON", "PARIS", "TOKYO", "BERLIN"],
        sports: ["SOCCER", "BASKETBALL", "TENNIS", "GOLF", "BASEBALL"],
        space: ["STAR", "PLANET", "MOON", "SUN", "GALAXY"],
        weather: ["RAIN", "SNOW", "SUN", "WIND", "STORM"]
    }
};

// Game state
let gameState = {
    grid: [],
    foundWords: new Set(),
    currentSelection: [],
    currentWord: "",
    score: 0,
    themeWords: [],
    themeWordsFound: 0,
    hintsUsed: 0,
    hintsAvailable: 0,
    hintWordsFound: 0,
    gameStarted: false,
    gameWon: false
};

// DOM elements
let boardElement = document.getElementById('board');
let currentWordElement = document.getElementById('current-word');
let themeDisplayElement = document.getElementById('theme-display');
let foundWordsListElement = document.getElementById('found-words-list');
let themeCountElement = document.getElementById('theme-count');
let completionElement = document.getElementById('completion');
let scoreElement = document.getElementById('score-display');
let hintCountElement = document.getElementById('hint-count');
let resetButton = document.getElementById('reset-btn');
let hintButton = document.getElementById('hint-btn');
let modal = document.getElementById('message-modal');
let modalTitle = document.getElementById('modal-title');
let modalMessage = document.getElementById('modal-message');
let modalClose = document.getElementById('modal-close');

// Initialize game
function initGame(theme) {
    if (!theme) {
        const themeNames = Object.keys(CONFIG.themes);
        theme = themeNames[Math.floor(Math.random() * themeNames.length)];
    }
    
    gameState.themeWords = CONFIG.themes[theme];
    gameState.themeWordsFound = 0;
    gameState.foundWords = new Set();
    gameState.currentSelection = [];
    gameState.currentWord = "";
    gameState.score = 0;
    gameState.hintsUsed = 0;
    gameState.hintWordsFound = 0;
    gameState.gameStarted = true;
    gameState.gameWon = false;
    
    generateGrid();
    renderBoard();
    updateUI();
    showTheme(theme);
}

// Generate grid with theme words
function generateGrid() {
    gameState.grid = [];
    
    // Initialize empty grid
    for (let row = 0; row < CONFIG.rows; row++) {
        gameState.grid[row] = [];
        for (let col = 0; col < CONFIG.cols; col++) {
            gameState.grid[row][col] = '';
        }
    }
    
    // Place theme words (simplified placement logic)
    let placedWords = [];
    let attempts = 0;
    
    while (placedWords.length < gameState.themeWords.length && attempts < 100) {
        const word = gameState.themeWords[placedWords.length];
        if (placeWord(word)) {
            placedWords.push(word);
        }
        attempts++;
    }
    
    // Fill remaining cells with random letters
    fillRemaining();
}

// Try to place a word on the grid
function placeWord(word) {
    const directions = [
        { dr: 0, dc: 1 },   // horizontal
        { dr: 1, dc: 0 },   // vertical
        { dr: 1, dc: 1 },   // diagonal down-right
        { dr: 1, dc: -1 }   // diagonal down-left
    ];
    
    const maxAttempts = 50;
    
    for (let attempt = 0; attempt < maxAttempts; attempt++) {
        const direction = directions[Math.floor(Math.random() * directions.length)];
        const startRow = Math.floor(Math.random() * CONFIG.rows);
        const startCol = Math.floor(Math.random() * CONFIG.cols);
        
        // Check if word fits
        let canPlace = true;
        for (let i = 0; i < word.length; i++) {
            const row = startRow + i * direction.dr;
            const col = startCol + i * direction.dc;
            
            if (row >= CONFIG.rows || col >= CONFIG.cols || col < 0) {
                canPlace = false;
                break;
            }
            
            // Check if cell is empty or same letter
            if (gameState.grid[row][col] !== '' && gameState.grid[row][col] !== word[i]) {
                canPlace = false;
                break;
            }
        }
        
        if (canPlace) {
            // Place the word
            for (let i = 0; i < word.length; i++) {
                const row = startRow + i * direction.dr;
                const col = startCol + i * direction.dc;
                gameState.grid[row][col] = word[i];
            }
            return true;
        }
    }
    
    return false;
}

// Fill remaining cells with random letters
function fillRemaining() {
    const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    
    for (let row = 0; row < CONFIG.rows; row++) {
        for (let col = 0; col < CONFIG.cols; col++) {
            if (gameState.grid[row][col] === '') {
                gameState.grid[row][col] = letters[Math.floor(Math.random() * letters.length)];
            }
        }
    }
}

// Render the board
function renderBoard() {
    boardElement.innerHTML = '';
    
    for (let row = 0; row < CONFIG.rows; row++) {
        for (let col = 0; col < CONFIG.cols; col++) {
            const cell = document.createElement('div');
            cell.className = 'cell';
            cell.textContent = gameState.grid[row][col];
            cell.dataset.row = row;
            cell.dataset.col = col;
            
            // Check if cell is selected
            if (isCellSelected(row, col)) {
                cell.classList.add('selected');
            }
            
            // Check if cell is used by a found word
            if (isCellUsed(row, col)) {
                cell.classList.add('used');
            }
            
            // Check if cell is part of a theme word
            if (isCellTheme(row, col)) {
                cell.classList.add('theme');
            }
            
            cell.addEventListener('click', () => handleCellClick(row, col));
            cell.addEventListener('mouseover', () => handleCellHover(row, col));
            cell.addEventListener('mouseout', handleCellOut);
            
            boardElement.appendChild(cell);
        }
    }
}

// Check if a cell is selected
function isCellSelected(row, col) {
    return gameState.currentSelection.some(
        pos => pos.row === row && pos.col === col
    );
}

// Check if a cell is used by a found word
function isCellUsed(row, col) {
    for (const word of gameState.foundWords) {
        const path = getWordPath(word);
        if (path.some(pos => pos.row === row && pos.col === col)) {
            return true;
        }
    }
    return false;
}

// Check if a cell is part of a theme word
function isCellTheme(row, col) {
    for (const word of gameState.themeWords) {
        if (gameState.foundWords.has(word)) {
            const path = getWordPath(word);
            if (path.some(pos => pos.row === row && pos.col === col)) {
                return true;
            }
        }
    }
    return false;
}

// Get word path (simplified - in real implementation, would store paths)
function getWordPath(word) {
    // This would be implemented to track actual paths
    return [];
}

// Handle cell click
function handleCellClick(row, col) {
    if (!gameState.gameStarted || gameState.gameWon) return;
    
    // If cell is used by another word, clear selection
    if (isCellUsed(row, col) && !isCellSelected(row, col)) {
        clearSelection();
        return;
    }
    
    // Toggle selection
    const index = gameState.currentSelection.findIndex(
        pos => pos.row === row && pos.col === col
    );
    
    if (index >= 0) {
        // Remove cell from selection
        gameState.currentSelection.splice(index, 1);
    } else {
        // Add cell to selection
        // Check if it's adjacent to the last selected cell
        if (gameState.currentSelection.length > 0) {
            const last = gameState.currentSelection[gameState.currentSelection.length - 1];
            const dr = Math.abs(row - last.row);
            const dc = Math.abs(col - last.col);
            
            // Must be adjacent (including diagonally)
            if (dr > 1 || dc > 1 || (dr === 0 && dc === 0)) {
                return; // Not adjacent
            }
        }
        
        gameState.currentSelection.push({ row, col });
    }
    
    updateCurrentWord();
    renderBoard();
}

// Handle cell hover
function handleCellHover(row, col) {
    // Could implement hover effects here
}

// Handle cell mouse out
function handleCellOut() {
    // Could implement hover effects here
}

// Update current word display
function updateCurrentWord() {
    gameState.currentWord = gameState.currentSelection
        .map(pos => gameState.grid[pos.row][pos.col])
        .join('');
    
    currentWordElement.textContent = gameState.currentWord || 'Select cells to form words';
}

// Clear current selection
function clearSelection() {
    gameState.currentSelection = [];
    gameState.currentWord = '';
    updateCurrentWord();
    renderBoard();
}

// Handle word submission
function submitWord() {
    const word = gameState.currentWord;
    
    if (word.length < 4) {
        showMessage('Words must be at least 4 letters long');
        return;
    }
    
    if (gameState.foundWords.has(word)) {
        showMessage('You already found this word!');
        return;
    }
    
    // Check if word is in the theme
    if (gameState.themeWords.includes(word)) {
        // Valid theme word
        gameState.foundWords.add(word);
        gameState.themeWordsFound++;
        gameState.score += word.length;
        updateHints();
        showMessage(`Found theme word: ${word}!`);
    } else {
        // Check if it's a valid English word (simplified)
        showMessage('Not a valid word in this puzzle');
        return;
    }
    
    clearSelection();
    renderBoard();
    updateUI();
    
    // Check if game is won
    if (gameState.themeWordsFound === gameState.themeWords.length) {
        gameState.gameWon = true;
        showMessage('Congratulations! You found all theme words!', true);
    }
}

// Update hints system
function updateHints() {
    gameState.hintWordsFound++;
    if (gameState.hintWordsFound % 3 === 0) {
        gameState.hintsAvailable++;
        hintCountElement.textContent = gameState.hintsAvailable;
    }
}

// Request hint
function requestHint() {
    if (gameState.hintsAvailable > 0) {
        gameState.hintsAvailable--;
        hintCountElement.textContent = gameState.hintsAvailable;
        
        // Find a theme word not yet found
        const remaining = gameState.themeWords.filter(
            word => !gameState.foundWords.has(word)
        );
        
        if (remaining.length > 0) {
            showMessage(`Hint: Find ${remaining[0]}!`);
        } else {
            showMessage('No hints available - you found all theme words!');
        }
    } else {
        showMessage('No hints available yet. Find more non-theme words!');
    }
}

// Update UI elements
function updateUI() {
    scoreElement.textContent = gameState.score;
    themeCountElement.textContent = `${gameState.themeWordsFound}/${gameState.themeWords.length}`;
    
    // Update completion percentage
    const totalCells = CONFIG.rows * CONFIG.cols;
    const usedCells = gameState.foundWords.size * 5; // Approximate
    const completion = Math.min(100, Math.round((usedCells / totalCells) * 100));
    completionElement.textContent = `${completion}%`;
    
    // Update found words list
    updateFoundWordsList();
}

// Update found words list
function updateFoundWordsList() {
    foundWordsListElement.innerHTML = '';
    
    for (const word of gameState.foundWords) {
        const item = document.createElement('div');
        item.className = 'word-item';
        
        if (gameState.themeWords.includes(word)) {
            item.classList.add('theme-word');
        }
        
        const wordSpan = document.createElement('span');
        wordSpan.className = 'word';
        wordSpan.textContent = word;
        
        const countSpan = document.createElement('span');
        countSpan.className = 'count';
        countSpan.textContent = word.length;
        
        item.appendChild(wordSpan);
        item.appendChild(countSpan);
        foundWordsListElement.appendChild(item);
    }
}

// Show theme name
function showTheme(theme) {
    themeDisplayElement.textContent = theme.charAt(0).toUpperCase() + theme.slice(1);
}

// Show message modal
function showMessage(message, isWin = false) {
    modalTitle.textContent = isWin ? '🎉 You Won!' : 'Message';
    modalMessage.textContent = message;
    modal.classList.remove('hidden');
}

// Hide message modal
function hideModal() {
    modal.classList.add('hidden');
}

// Reset game
function resetGame() {
    initGame();
}

// Event listeners
resetButton.addEventListener('click', resetGame);
hintButton.addEventListener('click', requestHint);
modalClose.addEventListener('click', hideModal);

// Handle Enter key for submitting word
document.addEventListener('keydown', (e) => {
    if (e.key === 'Enter' && gameState.currentWord.length > 0) {
        submitWord();
    } else if (e.key === 'Backspace' && gameState.currentWord.length > 0) {
        if (gameState.currentSelection.length > 0) {
            gameState.currentSelection.pop();
            updateCurrentWord();
            renderBoard();
        }
    }
});

// Initialize on page load
window.addEventListener('DOMContentLoaded', () => {
    initGame();
});
